tests/test_main.py

```diff
@@ -0,0 +1,371 @@
+        crud_event_qs = CRUDEvent.objects.filter(
+            object_id=obj.id, content_type=ContentType.objects.get_for_model(obj)
+        )
+        obj.save()
+        assert crud_event_qs.count() == 1
+
+    def test_fake_update(self, model):
+        obj = model.objects.create()
+        crud_event_qs = CRUDEvent.objects.filter(
+            object_id=obj.id, content_type=ContentType.objects.get_for_model(obj)
+        )
+        obj.save()
+        assert crud_event_qs.count() == 2
+
+    def test_delete(self, model):
+        obj = model.objects.create()
+        crud_event_qs = CRUDEvent.objects.filter(
+            object_id=obj.id, content_type=ContentType.objects.get_for_model(obj)
+        )
+        assert crud_event_qs.count() == 1
+
+        obj_id = obj.pk
+        obj.delete()
+        crud_event_qs = CRUDEvent.objects.filter(
+            object_id=obj_id, content_type=ContentType.objects.get_for_model(obj)
+        )
+        assert crud_event_qs.count() == 2
+
+    @pytest.mark.usefixtures("_audit_logger")
+    def test_propagate_exceptions(self, model, settings):
+        with pytest.raises(ValueError, match="Test exception"):
+            model.objects.create()
+
+        settings.DJANGO_EASY_AUDIT_PROPAGATE_EXCEPTIONS = False
+        try:
+            model.objects.create()
+        except ValueError:
+            pytest.fail("Unexpected ValueError")
+
+
+class TestAuditUUIDModels(TestAuditModels):
+    @pytest.fixture
+    def model(self):
+        return UUIDModel
+
+    @pytest.fixture
+    def fk_model(self):
+        return UUIDForeignKeyModel
+
+    @pytest.fixture
+    def m2m_model(self):
+        return UUIDM2MModel
+
+
+class TestAuditBigIntModels(TestAuditModels):
+    @pytest.fixture
+    def model(self):
+        return BigIntModel
+
+    @pytest.fixture
+    def fk_model(self):
+        return BigIntForeignKeyModel
+
+    @pytest.fixture
+    def m2m_model(self):
+        return BigIntM2MModel
+
+
+@pytest.mark.django_db
+class TestMiddleware:
+    def test_middleware_logged_in(self, user, client, username, password):
+        client.login(username=username, password=password)
+        client.post(reverse("test_app:create-obj"))
+        assert Model.objects.count() == 1
+
+        obj = Model.objects.all().first()
+        crud_event = CRUDEvent.objects.filter(
+            object_id=obj.id, content_type=ContentType.objects.get_for_model(obj)
+        ).first()
+        assert crud_event.user == user
+
+    def test_middleware_not_logged_in(self, client):
+        create_obj_url = reverse("test_app:create-obj")
+        client.post(create_obj_url)
+        assert Model.objects.count() == 1
+
+        obj = Model.objects.all().first()
+        crud_event = CRUDEvent.objects.filter(
+            object_id=obj.id, content_type=ContentType.objects.get_for_model(obj)
+        ).first()
+        assert crud_event.user is None
+
+    def test_manual_set_user(self, django_user_model, username, email, password):
```

Comment:
`django_user_model` is just a shortcut and avoids having to import `User`: https://pytest-django.readthedocs.io/en/latest/helpers.html#django-user-model